OpenAPI generator should use the class path when resolving type references

When a class in the project has a field which type is defined in an external library, the `generate-openapi-spec` task fails:

```
[INFO] --- vaadin-connect-maven-plugin:1.0-SNAPSHOT:generate-openapi-spec (vaadin-connect-resources) @ base-starter-connect ---
[INFO] Parsing java files from [/Users/viktor/IdeaProjects/vaadin-services/base-starter-connect/src/main/java]
[ERROR] Unsolved symbol : PasswordEncoder
com.github.javaparser.resolution.UnsolvedSymbolException: Unsolved symbol : PasswordEncoder

```

caused by having a field of type `org.springframework.security.crypto.password.PasswordEncoder` in the `StarterOAuthConfiguration` class in the project:
```java
@Configuration
public class StarterOAuthConfiguration {
    private static final String TEST_LOGIN = "test_login";
    private static final String TEST_PASSWORD = "test_password";

    private final PasswordEncoder encoder;
}
```

Instead of failing, in this case the Open API generator should use the project class path to resolve type references.

--- a/vaadin-connect-demo/src/main/java/com/vaadin/connect/demo/DemoVaadinService.java
+++ b/vaadin-connect-demo/src/main/java/com/vaadin/connect/demo/DemoVaadinService.java
@@ -11,8 +11,10 @@ import java.util.Map;
 import java.util.Objects;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.Version;
 
 import com.vaadin.connect.VaadinService;
+import com.vaadin.connect.demo.account.BeanWithTypeFromDependencies;
 import com.vaadin.connect.oauth.AnonymousAllowed;
 
 @VaadinService
@@ -61,6 +63,20 @@ public class DemoVaadinService {
   @RolesAllowed("ROLE_USER")
   public int addOne(int number) {
     return number + 1;
+  }
+
+  /**
+   * This method is to make sure that the generator won't fail the build when
+   * resolving types from project's dependencies.
+   *
+   * @param version
+   *          a version object from dependencies
+   *
+   * @return {@code null}
+   */
+  @PermitAll
+  public BeanWithTypeFromDependencies testTypeResolverMethod(Version version) {
+    return null;
   }
 
   @PermitAll

--- a/vaadin-connect-demo/src/main/java/com/vaadin/connect/demo/account/BeanWithTypeFromDependencies.java
+++ b/vaadin-connect-demo/src/main/java/com/vaadin/connect/demo/account/BeanWithTypeFromDependencies.java
@@ -1,0 +1,28 @@
+/*
+ * Copyright 2000-2019 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.connect.demo.account;
+
+import com.fasterxml.jackson.core.Version;
+
+/**
+ * A test bean which has a field with type from project's dependencies. This
+ * class is to make sure that the generator won't fail the build when resolving
+ * types from project's dependencies.
+ */
+public class BeanWithTypeFromDependencies {
+  Version version;
+  Account account;
+}

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/OpenApiSpecGeneratorMojo.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/OpenApiSpecGeneratorMojo.java
@@ -16,8 +16,15 @@
 
 package com.vaadin.connect.plugin;
 
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.stream.Collectors;
 
@@ -48,15 +55,36 @@ public class OpenApiSpecGeneratorMojo extends VaadinConnectMojoBase {
   @Override
   public void execute() {
     try {
-      List<String> jarPaths = project.getCompileClasspathElements().stream()
-          .filter(s -> s.endsWith(".jar")).collect(Collectors.toList());
       List<Path> sourcesPaths = project.getCompileSourceRoots().stream()
           .map(Paths::get).collect(Collectors.toList());
-      new OpenApiSpecGenerator(readApplicationProperties()).generateOpenApiSpec(
-          sourcesPaths, jarPaths, openApiJsonFile.toPath());
+      URL[] urlsForClassLoader = getUrls();
+      try (
+          URLClassLoader classLoader = new URLClassLoader(urlsForClassLoader)) {
+        new OpenApiSpecGenerator(readApplicationProperties())
+            .generateOpenApiSpec(sourcesPaths, classLoader,
+                openApiJsonFile.toPath());
+      }
     } catch (DependencyResolutionRequiredException e) {
       throw new IllegalStateException(
           "All dependencies need to be resolved before running the OpenAPI spec generator. Please resolve the dependencies and try again.",
+          e);
+    } catch (IOException e) {
+      throw new UncheckedIOException(
+          "I/O error happens when closing project's URLClassLoader after generating OpenAPI spec.",
+          e);
+    }
+  }
+
+  private URL[] getUrls() throws DependencyResolutionRequiredException {
+    List<URL> pathUrls = new ArrayList<>();
+    try {
+      for (String mavenCompilePath : project.getCompileClasspathElements()) {
+        pathUrls.add(new File(mavenCompilePath).toURI().toURL());
+      }
+      return pathUrls.toArray(new URL[pathUrls.size()]);
+    } catch (MalformedURLException e) {
+      throw new IllegalStateException(
+          "Can't create URLs from project class paths for generating OpenAPI spec.",
           e);
     }
   }

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/OpenApiSpecGeneratorMojo.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/OpenApiSpecGeneratorMojo.java
@@ -55,7 +55,9 @@ public class OpenApiSpecGeneratorMojo extends VaadinConnectMojoBase {
       new OpenApiSpecGenerator(readApplicationProperties()).generateOpenApiSpec(
           sourcesPaths, jarPaths, openApiJsonFile.toPath());
     } catch (DependencyResolutionRequiredException e) {
-      throw new IllegalStateException(e);
+      throw new IllegalStateException(
+          "All dependencies need to be resolved before running the OpenAPI spec generator. Please resolve the dependencies and try again.",
+          e);
     }
   }
 }

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/OpenApiSpecGeneratorMojo.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/OpenApiSpecGeneratorMojo.java
@@ -16,12 +16,16 @@
 
 package com.vaadin.connect.plugin;
 
+import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.List;
 import java.util.stream.Collectors;
 
+import org.apache.maven.artifact.DependencyResolutionRequiredException;
 import org.apache.maven.plugins.annotations.LifecyclePhase;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.plugins.annotations.ResolutionScope;
 import org.apache.maven.project.MavenProject;
 
 import com.vaadin.connect.plugin.generator.OpenApiSpecGenerator;
@@ -35,7 +39,7 @@ import com.vaadin.connect.plugin.generator.OpenApiSpecGenerator;
  * @see <a href="https://github.com/OAI/OpenAPI-Specification">OpenAPI
  *      specification</a>
  */
-@Mojo(name = "generate-openapi-spec", defaultPhase = LifecyclePhase.PROCESS_SOURCES)
+@Mojo(name = "generate-openapi-spec", defaultPhase = LifecyclePhase.PROCESS_SOURCES, requiresDependencyResolution = ResolutionScope.COMPILE_PLUS_RUNTIME)
 public class OpenApiSpecGeneratorMojo extends VaadinConnectMojoBase {
 
   @Parameter(defaultValue = "${project}", readonly = true, required = true)
@@ -43,9 +47,15 @@ public class OpenApiSpecGeneratorMojo extends VaadinConnectMojoBase {
 
   @Override
   public void execute() {
-    new OpenApiSpecGenerator(readApplicationProperties()).generateOpenApiSpec(
-        project.getCompileSourceRoots().stream().map(Paths::get)
-            .collect(Collectors.toList()),
-        openApiJsonFile.toPath());
+    try {
+      List<String> jarPaths = project.getCompileClasspathElements().stream()
+          .filter(s -> s.endsWith(".jar")).collect(Collectors.toList());
+      List<Path> sourcesPaths = project.getCompileSourceRoots().stream()
+          .map(Paths::get).collect(Collectors.toList());
+      new OpenApiSpecGenerator(readApplicationProperties()).generateOpenApiSpec(
+          sourcesPaths, jarPaths, openApiJsonFile.toPath());
+    } catch (DependencyResolutionRequiredException e) {
+      throw new IllegalStateException(e);
+    }
   }
 }

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiParser.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiParser.java
@@ -19,8 +19,6 @@ import javax.annotation.security.DenyAll;
 import javax.annotation.security.PermitAll;
 import javax.annotation.security.RolesAllowed;
 
-import java.io.IOException;
-import java.io.UncheckedIOException;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -58,8 +56,8 @@ import com.github.javaparser.resolution.types.ResolvedPrimitiveType;
 import com.github.javaparser.resolution.types.ResolvedReferenceType;
 import com.github.javaparser.resolution.types.ResolvedType;
 import com.github.javaparser.symbolsolver.JavaSymbolSolver;
+import com.github.javaparser.symbolsolver.resolution.typesolvers.ClassLoaderTypeSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
-import com.github.javaparser.symbolsolver.resolution.typesolvers.JarTypeSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
 import com.github.javaparser.utils.Pair;
 import com.github.javaparser.utils.SourceRoot;
@@ -100,7 +98,6 @@ class OpenApiParser {
   private static final String VAADIN_CONNECT_JWT_SECURITY_SCHEME = "vaadin-connect-jwt";
 
   private List<Path> javaSourcePaths = new ArrayList<>();
-  private List<String> jarPaths = new ArrayList<>();
   private OpenApiConfiguration configuration;
   private Set<String> usedSchemas;
   private Map<String, String> servicesJavadoc;
@@ -108,6 +105,7 @@ class OpenApiParser {
   private Map<String, PathItem> pathItems;
   private OpenAPI openApiModel;
   private final VaadinServiceNameChecker serviceNameChecker = new VaadinServiceNameChecker();
+  private ClassLoader typeResolverClassLoader;
 
   void addSourcePath(Path sourcePath) {
     if (sourcePath == null) {
@@ -121,14 +119,14 @@ class OpenApiParser {
   }
 
   /**
-   * Adds a jar path information to use when resolving the type used in a
-   * project from that jar
-   * 
-   * @param jarPath
-   *          path of the jar
+   * Set project's class loader which is used for resolving types from that
+   * project.
+   *
+   * @param typeResolverClassLoader
+   *          the project's class loader for type resolving
    */
-  void addJarPath(String jarPath) {
-    this.jarPaths.add(jarPath);
+  void setTypeResolverClassLoader(ClassLoader typeResolverClassLoader) {
+    this.typeResolverClassLoader = typeResolverClassLoader;
   }
 
   void setOpenApiConfiguration(OpenApiConfiguration configuration) {
@@ -176,14 +174,9 @@ class OpenApiParser {
   private ParserConfiguration createParserConfiguration() {
     CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver(
         new ReflectionTypeSolver(false));
-    try {
-      for (String jarPath : jarPaths) {
-        combinedTypeSolver.add(new JarTypeSolver(jarPath));
-      }
-    } catch (IOException e) {
-      throw new UncheckedIOException(
-          "I/O error happens when reading jar files. Please make sure that dependencies' jar files are accessible.",
-          e);
+    if (typeResolverClassLoader != null) {
+      combinedTypeSolver
+          .add(new ClassLoaderTypeSolver(typeResolverClassLoader));
     }
     return new ParserConfiguration()
         .setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiParser.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiParser.java
@@ -19,6 +19,8 @@ import javax.annotation.security.DenyAll;
 import javax.annotation.security.PermitAll;
 import javax.annotation.security.RolesAllowed;
 
+import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -57,6 +59,7 @@ import com.github.javaparser.resolution.types.ResolvedReferenceType;
 import com.github.javaparser.resolution.types.ResolvedType;
 import com.github.javaparser.symbolsolver.JavaSymbolSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
+import com.github.javaparser.symbolsolver.resolution.typesolvers.JarTypeSolver;
 import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
 import com.github.javaparser.utils.Pair;
 import com.github.javaparser.utils.SourceRoot;
@@ -97,6 +100,7 @@ class OpenApiParser {
   private static final String VAADIN_CONNECT_JWT_SECURITY_SCHEME = "vaadin-connect-jwt";
 
   private List<Path> javaSourcePaths = new ArrayList<>();
+  private List<String> jarPaths = new ArrayList<>();
   private OpenApiConfiguration configuration;
   private Set<String> usedSchemas;
   private Map<String, String> servicesJavadoc;
@@ -114,6 +118,17 @@ class OpenApiParser {
       throw new IllegalArgumentException("Java source path doesn't exist");
     }
     this.javaSourcePaths.add(sourcePath);
+  }
+
+  /**
+   * Add jar paths to help the symbol resolver resolves types in jar
+   * dependencies.
+   * 
+   * @param jarPath
+   *          path of the jar
+   */
+  void addJarPath(String jarPath) {
+    this.jarPaths.add(jarPath);
   }
 
   void setOpenApiConfiguration(OpenApiConfiguration configuration) {
@@ -143,9 +158,8 @@ class OpenApiParser {
     usedSchemas = new HashSet<>();
     servicesJavadoc = new HashMap<>();
 
-    ParserConfiguration parserConfiguration = new ParserConfiguration()
-        .setSymbolResolver(new JavaSymbolSolver(
-            new CombinedTypeSolver(new ReflectionTypeSolver(false))));
+    ParserConfiguration parserConfiguration = createParserConfiguration();
+
     javaSourcePaths.stream()
         .map(path -> new SourceRoot(path, parserConfiguration))
         .forEach(this::parseSourceRoot);
@@ -157,6 +171,20 @@ class OpenApiParser {
       }
     }
     addTagsInformation();
+  }
+
+  private ParserConfiguration createParserConfiguration() {
+    CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver(
+        new ReflectionTypeSolver(false));
+    try {
+      for (String jarPath : jarPaths) {
+        combinedTypeSolver.add(new JarTypeSolver(jarPath));
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+    return new ParserConfiguration()
+        .setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));
   }
 
   private void parseSourceRoot(SourceRoot sourceRoot) {
@@ -274,9 +302,10 @@ class OpenApiParser {
       if (field.isTransient()) {
         continue;
       }
-      field.getVariables().forEach(variableDeclarator -> schema.addProperties(
-          variableDeclarator.getNameAsString(),
-          parseResolvedTypeToSchema(variableDeclarator.getType().resolve())));
+      field.getVariables()
+          .forEach(variableDeclarator -> schema.addProperties(
+              variableDeclarator.getNameAsString(),
+              parseTypeToSchema(variableDeclarator.getType())));
     }
     return schema;
   }
@@ -388,7 +417,7 @@ class OpenApiParser {
   private MediaType createReturnMediaType(MethodDeclaration methodDeclaration) {
     MediaType mediaItem = new MediaType();
     Type methodReturnType = methodDeclaration.getType();
-    mediaItem.schema(parseResolvedTypeToSchema(methodReturnType.resolve()));
+    mediaItem.schema(parseTypeToSchema(methodReturnType));
     return mediaItem;
   }
 
@@ -411,8 +440,7 @@ class OpenApiParser {
     Schema requestSchema = new ObjectSchema();
     requestBodyObject.schema(requestSchema);
     methodDeclaration.getParameters().forEach(parameter -> {
-      Schema paramSchema = parseResolvedTypeToSchema(
-          parameter.getType().resolve());
+      Schema paramSchema = parseTypeToSchema(parameter.getType());
       paramSchema
           .description(paramsDescription.get(parameter.getNameAsString()));
 
@@ -421,6 +449,17 @@ class OpenApiParser {
       requestSchema.addProperties(name, paramSchema);
     });
     return requestBody;
+  }
+
+  private Schema parseTypeToSchema(Type javaType) {
+    try {
+      return parseResolvedTypeToSchema(javaType.resolve());
+    } catch (Exception e) {
+      LoggerFactory.getLogger(OpenApiParser.class).info(String.format(
+          "Can't resolve type %s for creating custom OpenAPI Schema. Use the default ObjectSchema instead.",
+          javaType.asString()), e);
+    }
+    return new ObjectSchema();
   }
 
   private Schema parseResolvedTypeToSchema(ResolvedType resolvedType) {

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiParser.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiParser.java
@@ -121,8 +121,8 @@ class OpenApiParser {
   }
 
   /**
-   * Add jar paths to help the symbol resolver resolves types in jar
-   * dependencies.
+   * Adds a jar path information to use when resolving the type used in a
+   * project from that jar
    * 
    * @param jarPath
    *          path of the jar
@@ -181,7 +181,9 @@ class OpenApiParser {
         combinedTypeSolver.add(new JarTypeSolver(jarPath));
       }
     } catch (IOException e) {
-      throw new UncheckedIOException(e);
+      throw new UncheckedIOException(
+          "I/O error happens when reading jar files. Please make sure that dependencies' jar files are accessible.",
+          e);
     }
     return new ParserConfiguration()
         .setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));
@@ -456,7 +458,7 @@ class OpenApiParser {
       return parseResolvedTypeToSchema(javaType.resolve());
     } catch (Exception e) {
       LoggerFactory.getLogger(OpenApiParser.class).info(String.format(
-          "Can't resolve type %s for creating custom OpenAPI Schema. Use the default ObjectSchema instead.",
+          "Can't resolve type '%s' for creating custom OpenAPI Schema. Using the default ObjectSchema instead.",
           javaType.asString()), e);
     }
     return new ObjectSchema();

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiSpecGenerator.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiSpecGenerator.java
@@ -86,15 +86,14 @@ public class OpenApiSpecGenerator {
    *
    * @param sourcesPaths
    *          the source root to be analyzed
-   * @param jarPaths
-   *          jar paths information to use when resolving the types used in a
-   *          project from those jars
+   * @param classLoader
+   *          the ClassLoader which is able to load the classes in sourcesPaths
    * @param specOutputFile
    *          the target file to write the generation output to
    */
   public void generateOpenApiSpec(Collection<Path> sourcesPaths,
-      Collection<String> jarPaths, Path specOutputFile) {
-    jarPaths.forEach(generator::addJarPath);
+      ClassLoader classLoader, Path specOutputFile) {
+    generator.setTypeResolverClassLoader(classLoader);
     generateOpenApiSpec(sourcesPaths, specOutputFile);
   }
 

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiSpecGenerator.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiSpecGenerator.java
@@ -81,6 +81,23 @@ public class OpenApiSpecGenerator {
     GeneratorUtils.writeToFile(specOutputFile, Json.pretty(openAPI));
   }
 
+  /**
+   * Generates the open api spec file based on the sources provided.
+   *
+   * @param sourcesPaths
+   *          the source root to be analyzed
+   * @param jarPaths
+   *          the jar paths which help the parser resolves the types in those
+   *          jar dependencies
+   * @param specOutputFile
+   *          the target file to write the generation output to
+   */
+  public void generateOpenApiSpec(Collection<Path> sourcesPaths,
+      Collection<String> jarPaths, Path specOutputFile) {
+    jarPaths.forEach(generator::addJarPath);
+    generateOpenApiSpec(sourcesPaths, specOutputFile);
+  }
+
   private OpenApiConfiguration extractOpenApiConfiguration(
       Properties applicationProperties) {
     String endpoint = applicationProperties.getProperty(ENDPOINT,

--- a/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiSpecGenerator.java
+++ b/vaadin-connect-maven-plugin/src/main/java/com/vaadin/connect/plugin/generator/OpenApiSpecGenerator.java
@@ -64,7 +64,7 @@ public class OpenApiSpecGenerator {
   }
 
   /**
-   * Generates the open api spec file based on the sources provided.
+   * Generates the OpenAPI spec file based on the sources provided.
    *
    * @param sourcesPaths
    *          the source root to be analyzed
@@ -82,13 +82,13 @@ public class OpenApiSpecGenerator {
   }
 
   /**
-   * Generates the open api spec file based on the sources provided.
+   * Generates the OpenAPI spec file based on the sources provided.
    *
    * @param sourcesPaths
    *          the source root to be analyzed
    * @param jarPaths
-   *          the jar paths which help the parser resolves the types in those
-   *          jar dependencies
+   *          jar paths information to use when resolving the types used in a
+   *          project from those jars
    * @param specOutputFile
    *          the target file to write the generation output to
    */

--- a/vaadin-connect-maven-plugin/src/test/java/com/vaadin/connect/plugin/generator/collectionservice/CollectionTestService.java
+++ b/vaadin-connect-maven-plugin/src/test/java/com/vaadin/connect/plugin/generator/collectionservice/CollectionTestService.java
@@ -15,9 +15,9 @@
  */
 package com.vaadin.connect.plugin.generator.collectionservice;
 
-import com.vaadin.connect.VaadinService;
-
 import java.util.ArrayList;
+
+import com.vaadin.connect.VaadinService;
 
 @VaadinService
 public class CollectionTestService {

